File: examples/python/stackby_client.py
"""
Stackby API Client for Python

Usage:
    client = StackbyClient('your_api_key', 'stackId', 'tableId')
    rows = client.list_rows()
    client.create_row({'Name': 'John', 'Email': 'john@example.com'})
"""

import os
import requests
from typing import Union, List, Dict, Any


class StackbyClient:
    BASE_URL = 'https://stackby.com/api/betav1'

    def __init__(self, api_key: str, stack_id: str, table_id: str):
        self.api_key = api_key
        self.stack_id = stack_id
        self.table_id = table_id
        self.headers = {
            'api-key': api_key,
            'Content-Type': 'application/json'
        }

    def _request(self, method: str, endpoint: str, json: dict = None) -> Any:
        url = f'{self.BASE_URL}{endpoint}'
        response = requests.request(method, url, headers=self.headers, json=json)
        response.raise_for_status()
        return response.json()

    def list_rows(self) -> List[Dict]:
        """List all rows in the table."""
        return self._request('GET', f'/rowlist/{self.stack_id}/{self.table_id}')

    def create_rows(self, fields: Union[Dict, List[Dict]]) -> List[Dict]:
        """
        Create one or more rows.

        IMPORTANT: The API uses "field" (singular), NOT "fields" (plural).
        """
        if isinstance(fields, dict):
            fields = [fields]

        records = [{'field': f} for f in fields]

        return self._request(
            'POST',
            f'/rowcreate/{self.stack_id}/{self.table_id}',
            {'records': records}
        )

    def update_rows(self, updates: Union[Dict, List[Dict]]) -> List[Dict]:
        """
        Update one or more rows.

        Args:
            updates: {'id': 'rowId', 'fields': {...}} or list of updates
        """
        if isinstance(updates, dict):
            updates = [updates]

        records = [{'id': u['id'], 'field': u['fields']} for u in updates]

        return self._request(
            'PATCH',
            f'/rowupdate/{self.stack_id}/{self.table_id}',
            {'records': records}
        )

    def delete_rows(self, row_ids: Union[str, List[str]]) -> Dict:
        """
        Delete one or more rows.

        Note: Row IDs are passed as query parameters, not in the body.
        """
        if isinstance(row_ids, str):
            row_ids = [row_ids]

        query_string = '&'.join([f'rowIds[]={rid}' for rid in row_ids])
        url = f'{self.BASE_URL}/rowdelete/{self.stack_id}/{self.table_id}?{query_string}'

        response = requests.delete(url, headers={'api-key': self.api_key})
        response.raise_for_status()
        return response.json()


def main():
    """Example usage."""
    client = StackbyClient(
        api_key=os.environ['STACKBY_API_KEY'],
        stack_id=os.environ['STACKBY_STACK_ID'],
        table_id=os.environ['STACKBY_TABLE_ID']
    )

    # List rows
    rows = client.list_rows()
    print(f'Existing rows: {len(rows)}')

    # Create a row
    new_row = client.create_rows({
        'Name': 'Test User',
        'Email': 'test@example.com'
    })
    print(f'Created: {new_row[0]["id"]}')

    # Update the row
    client.update_rows({
        'id': new_row[0]['id'],
        'fields': {'Name': 'Updated User'}
    })
    print('Updated')

    # Delete the row
    client.delete_rows(new_row[0]['id'])
    print('Deleted')


if __name__ == '__main__':
    main()

File: examples/javascript/stackby-client.js
/**
 * Stackby API Client for JavaScript/Node.js
 *
 * Usage:
 *   const client = new StackbyClient('your_api_key', 'stackId', 'tableId');
 *   const rows = await client.listRows();
 *   await client.createRow({ Name: 'John', Email: 'john@example.com' });
 */

class StackbyClient {
  constructor(apiKey, stackId, tableId) {
    this.apiKey = apiKey;
    this.stackId = stackId;
    this.tableId = tableId;
    this.baseUrl = 'https://stackby.com/api/betav1';
  }

  async request(method, endpoint, body = null) {
    const url = `${this.baseUrl}${endpoint}`;

    const options = {
      method,
      headers: {
        'api-key': this.apiKey,
        'Content-Type': 'application/json',
      },
    };

    if (body) {
      options.body = JSON.stringify(body);
    }

    const response = await fetch(url, options);

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`Stackby API error ${response.status}: ${error}`);
    }

    return response.json();
  }

  /**
   * List all rows in the table
   */
  async listRows() {
    return this.request('GET', `/rowlist/${this.stackId}/${this.tableId}`);
  }

  /**
   * Create one or more rows
   * @param {Object|Object[]} fields - Single row fields or array of row fields
   */
  async createRows(fields) {
    const records = Array.isArray(fields)
      ? fields.map(f => ({ field: f }))
      : [{ field: fields }];

    // IMPORTANT: Use "field" (singular), NOT "fields"
    return this.request('POST', `/rowcreate/${this.stackId}/${this.tableId}`, {
      records
    });
  }

  /**
   * Update one or more rows
   * @param {Object|Object[]} updates - { id, fields } or array of updates
   */
  async updateRows(updates) {
    const records = Array.isArray(updates)
      ? updates.map(u => ({ id: u.id, field: u.fields }))
      : [{ id: updates.id, field: updates.fields }];

    return this.request('PATCH', `/rowupdate/${this.stackId}/${this.tableId}`, {
      records
    });
  }

  /**
   * Delete one or more rows
   * @param {string|string[]} rowIds - Single row ID or array of IDs
   */
  async deleteRows(rowIds) {
    const ids = Array.isArray(rowIds) ? rowIds : [rowIds];
    const queryString = ids.map(id => `rowIds[]=${id}`).join('&');

    const url = `${this.baseUrl}/rowdelete/${this.stackId}/${this.tableId}?${queryString}`;

    const response = await fetch(url, {
      method: 'DELETE',
      headers: { 'api-key': this.apiKey },
    });

    return response.json();
  }
}

// Example usage
async function main() {
  const client = new StackbyClient(
    process.env.STACKBY_API_KEY,
    process.env.STACKBY_STACK_ID,
    process.env.STACKBY_TABLE_ID
  );

  // List rows
  const rows = await client.listRows();
  console.log('Existing rows:', rows.length);

  // Create a row
  const newRow = await client.createRows({
    Name: 'Test User',
    Email: 'test@example.com'
  });
  console.log('Created:', newRow[0].id);

  // Update the row
  await client.updateRows({
    id: newRow[0].id,
    fields: { Name: 'Updated User' }
  });
  console.log('Updated');

  // Delete the row
  await client.deleteRows(newRow[0].id);
  console.log('Deleted');
}

// Run if executed directly
if (typeof require !== 'undefined' && require.main === module) {
  main().catch(console.error);
}

module.exports = { StackbyClient };
